src/main/java/org/opensearch/neuralsearch/query/HybridQuery.java

```diff
@@ -106,22 +111,40 @@ public Query rewrite(IndexSearcher indexSearcher) throws IOException {
             return new MatchNoDocsQuery("empty HybridQuery");
         }
 
-        boolean actuallyRewritten = false;
-        List<Query> rewrittenSubQueries = new ArrayList<>();
+        HybridQueryRewriteCollectorManager manager = new HybridQueryRewriteCollectorManager(indexSearcher);
+        List<Callable<Void>> queryRewriteTasks = new ArrayList<>();
+        List<HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>>> collectors = new ArrayList<>();
         for (Query subQuery : subQueries) {
-            Query rewrittenSub = subQuery.rewrite(indexSearcher);
-            /* we keep rewrite sub-query unless it's not equal to itself, it may take multiple levels of recursive calls
-               queries need to be rewritten from high-level clauses into lower-level clauses because low-level clauses
-               perform better. For hybrid query we need to track progress of re-write for all sub-queries */
-            actuallyRewritten |= rewrittenSub != subQuery;
-            rewrittenSubQueries.add(rewrittenSub);
+            HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>> collector = manager.newCollector();
+            collectors.add(collector);
+            queryRewriteTasks.add(() -> rewriteQuery(subQuery, collector));
         }
 
-        if (actuallyRewritten) {
-            return new HybridQuery(rewrittenSubQueries);
+        HybridQueryExecutor.getExecutor().invokeAll(queryRewriteTasks);
+
+        final boolean anyQueryRewrite = manager.anyQueryRewrite(collectors);
+        if (anyQueryRewrite == false) {
+            return super.rewrite(indexSearcher);
         }
+        final List<Query> rewrittenSubQueries = manager.getRewriteQueries(collectors);
+        return new HybridQuery(rewrittenSubQueries);
+    }
 
-        return super.rewrite(indexSearcher);
+    private Void rewriteQuery(Query query, HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>> collector) {
+        collector.collect(indexSearcher -> {
+            try {
+                Query rewrittenQuery = query.rewrite(indexSearcher);
+                /* we keep rewrite sub-query unless it's not equal to itself, it may take multiple levels of recursive calls
+                queries need to be rewritten from high-level clauses into lower-level clauses because low-level clauses
+                perform better. For hybrid query we need to track progress of re-write for all sub-queries */
+
+                boolean actuallyRewritten = rewrittenQuery != query;
+                return new AbstractMap.SimpleEntry(rewrittenQuery, actuallyRewritten);
```

Comment:
can we avoid storing queries in a map? and who is going to store this simple map entry?

Reply:
HybridQueryExecutorCollector.collect() will store this entry. If we don't store map, we can't assemble back to caller once the execution is completed.

Reply:
You can use optional instead.
```
return rewrittenQuery != query ? Optional.empty() : Optional.of(rewrittenQuery)
```

Reply:
and +1 to Navneet's point regarding map and specifically using Query as a key. I was using similar approach in initial implementation of hybrid query, problem is that for some queries `hashCode` can be slow. 
How much time this will be called - one time per sub-query x num of shards?

Reply:
Used optional to avoid null condition, but we need query irrespective of whether query is re-written or not.

Reply:
@martin-gaievski We are using list to store the result. I don't believe hashcode will be calculated while adding it to list.

Reply:
understood, I mismatch the Map.Entry<Query, Boolean> with Map<Query, Boolean>, I guess the Map.Entry is used to operate on pair of objects. I'm good now, thanks Vijay

Reply:
> Used optional to avoid null condition, but we need query irrespective of whether query is re-written or not.

We can explicitly check for null, my point was performance. Doing first `Optional.of(rewrittenQuery)`, and later `optional.get()` both will add some cycles of wrapping query object into Optional and then unwrapping it. If we have done benchmarks with Optional and without it then we can use it as datapoint to make a decision.

Reply:
Sounds good. I will run benchmark with and without before merging into main. In the meantime, can we keep Optional? What do you think?

Reply:
Yes, I'm not sure what is the delta in performance if any, let's keep it with Optional, run benchmark and act per per data

Reply:
the intent was never to have an Optional as a member variable, the intent was to only return optional to indicate the value is not rewritten. It got misunderstood.

I realize that the query is needed irrespective of if its rewritten so Optional cannot be used. An `Map.Entry` might be a good option here but will leave it up to you.